Add optional is_coalesced argument to sparse coo tensor factory function.

pearu opened the pull request:
Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced=is_coalesced)
```

All sparse coo factory functions that take indices as input support the `is_coalesced` argument.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107903
* #107150
* #107777
* __->__ #107638

pearu committed:
Add optional is_coalesced argument to sparse coo tensor factory function.

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6821,11 +6823,13 @@
   autogen: _sparse_coo_tensor_with_dims.out
 
-- func: _sparse_coo_tensor_with_dims_and_tensors(int sparse_dim, int dense_dim, SymInt[] size, Tensor indices, Tensor values, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=False) -> Tensor
+- func: _sparse_coo_tensor_with_dims_and_tensors(int sparse_dim, int dense_dim, SymInt[] size, Tensor indices, Tensor values, bool is_coalesced, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=False) -> Tensor
```
Why not kwarg only?

pearu replied:
Indeed, making `is_coalesced` a keyword argument resulted in a better code (less test failures, less code).

albanD reviewed (commented):
Sounds good to me. Will let @amjames do a final review from the sparse side

pearu committed:
Update on "Add optional is_coalesced argument to sparse coo tensor factory function."

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced)
```

Note that `is_coalesced` argument support is implemented only for sparse coo factory function variant that explicitly specifies indices, values, and the shape of the tensor.




[ghstack-poisoned]

pearu committed:
Update on "Add optional is_coalesced argument to sparse coo tensor factory function."

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced)
```

Note that `is_coalesced` argument support is implemented only for sparse coo factory function variant that explicitly specifies indices, values, and the shape of the tensor.




[ghstack-poisoned]

pearu committed:
Update on "Add optional is_coalesced argument to sparse coo tensor factory function."

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced)
```

Note that `is_coalesced` argument support is implemented only for sparse coo factory function variant that explicitly specifies indices, values, and the shape of the tensor.




[ghstack-poisoned]

pearu committed:
Update on "Add optional is_coalesced argument to sparse coo tensor factory function."

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced)
```

~Note that `is_coalesced` argument support is implemented only for sparse coo factory function variant that explicitly specifies indices, values, and the shape of the tensor.~
UPDATE: all sparse coo factory functions that take indices as input support the `is_coalesced` argument.




[ghstack-poisoned]

amjames commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -204,6 +206,20 @@ SparseTensor new_with_dims_and_tensor_sparse_symint(
+  // likely that these methods are overly restrictive when estimating
+  // is_coalesced state. The condition `!is_coalesced && self._nnz() <
+  // 2` provides a way to detect and optimize such methods with
+  // respect to estimating the is_coalesced state.
```
A thought: OOS for this change, but relevant to this TODO

Any method that currents flips coalesced to `false` could inherit the `nnz < 2` optimization if the logic were placed within `SparseTensorImpl::set_coalesced` instead of `SparseTensorImpl::set_nnz_and_narrow` where I originally put it. 

That could be confusing and/or dangerous though as `impl->set_coalesced(false)` would become silently equivalent to `impl->set_coalesed(nnz < 2)`

pearu replied:
In addition, I suggest that the following invariant always holds:
```
sparse_coo_tensor(..., is_coalesced=is_coalesced).is_coalesced() == is_coalesced
```
including tensors that have `nnz < 2`.

This property is useful when explicitly testing code pathways that operate on uncoalesced  tensors.

amjames commented on torch/_torch_docs.py:
```diff
@@ -10549,6 +10551,13 @@ def merge_dicts(*dicts):
+        coalesced tensor that is a prerequisite for various algoritms
+        that process tensor indices to produce expected results.
+        Default: if None, the coalesced flag may be estimated using
+        the conservative value ``False`` as an initial state.
```
We should explicitly note the danger zone here, something like: 

```
If the `check_invariants` flag is false no error will be raised if the prerequisites are not met and this will likely lead to silently incorrect results. 
```

Should cover it

pearu replied:
Applied, thanks!

pearu committed:
Update on "Add optional is_coalesced argument to sparse coo tensor factory function."

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced)
```

All sparse coo factory functions that take indices as input support the `is_coalesced` argument.




[ghstack-poisoned]

pearu committed:
Update on "Add optional is_coalesced argument to sparse coo tensor factory function."

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced)
```

All sparse coo factory functions that take indices as input support the `is_coalesced` argument.




[ghstack-poisoned]

cpuhrsch commented on torch/_torch_docs.py:
```diff
@@ -10549,6 +10551,16 @@ def merge_dicts(*dicts):
+    is_coalesced (bool, optional): When specified, set the coalesced
+        flag to :attr:`is_coalesced`. If ``True``, the caller is
+        responsible for providing tensor indices that correspond to a
+        coalesced tensor that is a prerequisite for various algoritms
```
nit: algorithms

Also, I think we can simplify the documentation a bit.

```
    is_coalesced (bool, optional): When``True``, the caller is
        responsible for providing tensor indices that correspond to a
        coalesced tensor.  If the :attr:`check_invariants` flag is False,
        no error will be raised if the prerequisites are not met and this will
        lead to silently incorrect results.
        Default: None, to save on overhead the resulting Tensor will have
        the coalesced flag set to ``False``.
```

For the default case, I think we're always returning an uncoalesced COO Tensor when using this constructor? Maybe we don't need None, False, True but can just set `is_coalesced=False` by default?

pearu replied:
> For the default case, I think we're always returning an uncoalesced COO Tensor when using this constructor?

No: Currently, COO tensors with `nnz < 2`, are typically returned as coalesced tensors while there exists a few untypical cases where `nnz < 2` condition is explicitly (and unnecessarily, imho) overridden. So, the `is_coalesced=None` corresponds to the current behavior where the coalesce state is estimated from the shape information of input tensors. However, when `is_coalesced` is not `None`, the coalesce state will be forced according to the user-specified value. So, having `is_coalesced` with three possible values (`None`, `False`, and `True`) still makes sense to me: using `None` hints that the coalesce state should be estimated using the conservative initial value `False`, and the other two values force the coalesce state as specified.

cpuhrsch replied:
Ok, then I'd suggest to write the following documentation to include this.

```
    is_coalesced (bool, optional): When``True``, the caller is
        responsible for providing tensor indices that correspond to a
        coalesced tensor.  If the :attr:`check_invariants` flag is False,
        no error will be raised if the prerequisites are not met and this will
        lead to silently incorrect results. To force coalescion please use
        `coalesce`(<reference>) on the resulting Tensor.
        Default: None: except for trivial cases (e.g. nnz < 2) the resulting Tensor
        has is_coalesced set to ``False```.
```

pearu replied:
Applied the suggestion.

cpuhrsch reviewed (changes requested):
See inline comment

pearu committed:
Update on "Add optional is_coalesced argument to sparse coo tensor factory function."

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced=s_coalesced)
```

All sparse coo factory functions that take indices as input support the `is_coalesced` argument.




[ghstack-poisoned]

cpuhrsch commented on torch/_torch_docs.py:
```diff
@@ -10549,6 +10551,13 @@ def merge_dicts(*dicts):
+        False, no error will be raised if the prerequisites are not
+        met and this will lead to silently incorrect results.
+        Default: None, to save on overhead the resulting Tensor will
+        have the coalesced flag set to ``False``.
```
But then it sounds like None always means False. You just pointed out that None can also sometimes mean True.

pearu replied:
@cpuhrsch, sorry, something went wrong in applying your suggestion last night... Now fixed, please review.

pearu committed:
Update on "Add optional is_coalesced argument to sparse coo tensor factory function."

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced=is_coalesced)
```

All sparse coo factory functions that take indices as input support the `is_coalesced` argument.




[ghstack-poisoned]
